### PR TITLE
PA-2158 Fixed SPL Report Baseline Integrity

### PR DESCRIPTION
**Pull request by Fosol**
Originally the project import would generate an SPL Report with snapshots to represent a historical report.  This update makes that optional when importing projects and includes a new endpoint that allows for manually submitting a full SPL Report (including snapshots).

Additionally Disposal Projects will now auto-calculate Net Proceeds when values change on the Project Information tab.

You should now be able to run and debug the API locally within VSCode without having to open up a separate instance of VSCode.

- Updated Project Import
- Fixed VSCode Launch API

### CODE REVIEW

File: backend/api/Areas/Project/Controllers/ReportController.cs [hunk 1]
```diff
@@ -126,7 +128,16 @@ public IActionResult GetProjectReportSnapshots(int reportId)
         [SwaggerOperation(Tags = new[] { "project" })]
         public IActionResult AddProjectReport(ProjectReportModel model)
         {
-            var report = _pimsService.ProjectReport.Add(_mapper.Map<Entity.ProjectReport>(model));
+            Entity.ProjectReport report;
+
+            if (model.Snapshots?.Any() == true)
```
Comment (Fosol) [inline backend/api/Areas/Project/Controllers/ReportController.cs:133]: If snapshots are provided, create a report with them.

File: backend/dal/Configuration/ProjectSnapshotConfiguration.cs [hunk 2]
```diff
@@ -15,7 +15,6 @@ public override void Configure(EntityTypeBuilder<ProjectSnapshot> builder)
             builder.ToTable("ProjectSnapshots");
 
             builder.HasKey(m => m.Id);
-            builder.Property(m => m.Id).IsRequired();
```
Comment (Fosol) [inline backend/dal/Configuration/ProjectSnapshotConfiguration.cs:18]: An attempt to resolve the Entity Framework Bug, but for some reason it still wouldn't accept the auto generated entities from the mapper.

File: backend/dal/Services/Concrete/ProjectReportService.cs [hunk 3]
```diff
@@ -241,10 +241,51 @@ public void Remove(ProjectReport report)
                 .SingleOrDefault(p => p.Id == report.Id) ?? throw new KeyNotFoundException();
             if (originalReport.IsFinal) throw new InvalidOperationException($"Unable to delete FINAL project reports.");
 
+            // Remove all snapshots associated with this report.
+            var snapshots = this.Context.ProjectSnapshots.Where(s => s.SnapshotOn == originalReport.To.Value).ToArray();
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ProjectReportService.cs:245]: We were leaving orphaned snapshots previously.

File: frontend/src/features/projects/common/components/ProjectFinancialTable.tsx [hunk 4]
```diff
@@ -14,6 +20,32 @@ export interface IProjectFinacialTableProps {
  * Provides a way to display and edit project financial information.
  */
 export const ProjectFinancialTable = ({ disabled, title }: IProjectFinacialTableProps) => {
+  const { values, setFieldValue } = useFormikContext<IProject>();
+
+  const market = getIn(values, 'market');
+  const interestComponent = getIn(values, 'interestComponent');
+  const salesCost = getIn(values, 'salesCost');
+  const netBook = getIn(values, 'netBook');
+  React.useEffect(() => {
```
Comment (Fosol) [inline frontend/src/features/projects/common/components/ProjectFinancialTable.tsx:29]: Calculate and update the values on the Project Information tab.

File: backend/tests/unit/api/Controllers/Tools/ImportControllerTest.cs [hunk 5]
```diff
@@ -196,7 +197,7 @@ public void ImportProjects_Success()
             };
 
             // Act
-            var result = controller.ImportProjects(projects, false);
+            var result = controller.ImportProjects(projects, false, DateTime.UtcNow);
```
Comment (agalbrai) [inline backend/tests/unit/api/Controllers/Tools/ImportControllerTest.cs:200]: just curious, why UTC time?

**Review by agalbrai — COMMENTED** [review]
What are all the empty files for?

File: backend/tests/unit/api/Controllers/Tools/ImportControllerTest.cs (hunk 5 above)
Comment (Fosol) [reply backend/tests/unit/api/Controllers/Tools/ImportControllerTest.cs:200]: We always store dates in the DB with a single standard time.  The convention is UTC.  If you store different timezone values in the DB you then also need to store what timezone you stored the date.  Which historically was problematic and unreliable.

**Comment by Fosol** [comment]
There are technically no empty files, but rather the files are identified as executable files for linux based systems.

File: backend/api/Areas/Project/Controllers/ReportController.cs (hunk 1 above)
Comment (devinleighsmith) [reply backend/api/Areas/Project/Controllers/ReportController.cs:133]: is there a need for == true?

File: backend/api/Areas/Tools/Helpers/ImportProjectsHelper.cs [hunk 6]
```diff
@@ -263,7 +263,8 @@ private Entity.Project Merge(Entity.Project project, Model.ImportProjectModel mo
 
             // A prior net proceeds was provided, which means a prior snapshot needs to be generated.
             // If the project already exists, don't add prior snapshots.
-            if (model.PriorNetProceeds.HasValue && project.Id == 0)
+            // Only create snapshots if a `snapshotOn` date has been provied.
```
Comment (devinleighsmith) [inline backend/api/Areas/Tools/Helpers/ImportProjectsHelper.cs:266]: provided

File: backend/dal/Services/Concrete/ProjectReportService.cs [hunk 7]
```diff
@@ -241,10 +241,51 @@ public void Remove(ProjectReport report)
...
+        /// This provides a way to manually generate snapshots with specific projects and historical values.
+        /// Requires a valid 'report.To' date.
+        /// </summary>
+        /// <param name="report"></param>
+        /// <param name="snapshots"></param>
+        public ProjectReport Add(ProjectReport report, IEnumerable<ProjectSnapshot> snapshots)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.ReportsSpl);
+            if (!report.To.HasValue) throw new InvalidOperationException("Argument 'report.To' must have a valid date.");
+            if (report.From.HasValue && report.From > report.To) throw new InvalidOperationException("Argument 'report.From' must be equal to or greater than 'report.To'.");
```
Comment (devinleighsmith) [inline backend/dal/Services/Concrete/ProjectReportService.cs:263]: I think its equal to or less than

**Comment by devinleighsmith** [comment]
There is a whole bunch of empty .sh files at the end of this PR, what's that all about?

**Review by devinleighsmith — APPROVED** [review]

**Comment by Fosol** [comment]
See above comment about making the files executable

File: backend/dal/Services/Concrete/ProjectReportService.cs (hunk 7 above)
Comment (Fosol) [reply backend/dal/Services/Concrete/ProjectReportService.cs:263]: If the date is inverse (from > to) then throw error.

**Pull request merged**